{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counting DNA Nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'250 242 222 231'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def counting_nucleotides(dna_string):\n",
    "    a = 0\n",
    "    c = 0\n",
    "    g = 0\n",
    "    t = 0\n",
    "    for x in dna_string:\n",
    "        if x.upper() == 'A':\n",
    "            a += 1\n",
    "        elif x.upper() == 'C':\n",
    "            c += 1\n",
    "        elif x.upper() == 'G':\n",
    "            g += 1\n",
    "        elif x.upper() == 'T':\n",
    "            t += 1\n",
    "    return f\"{a} {c} {g} {t}\"\n",
    "\n",
    "counting_nucleotides('TATGTCGCGGCAACTGGTCCCATGAGTCACACCTATCTGCGCCTCTGGGGAGAGTATTCTCAACATGCTCGCTTTTGAAGCCAAACCTTACTATTACGAGACCTACGAAGGGAAATGGAATAGTACACCAACACGGAGCGCATGGAAGCGCTCTCCGCGGTTGCTCATGTCAAATAGCTGGCAGGTTCATAGAGGGGCGATGGTAGACTGAAAACATCAGTTTCAGTAATCGCTTGGTCCGCTACACAGCACACTCTACGGATATTGAAGTCCGTATAATACCAAACCAGTCAGCCAACGAGAAGAGTAGTCTTCAGCGATTAGGTACGCCCTTTTAATGTGGGCAGCCAGTGCGGTAGGCCGCTATACTGTACTGCATACCGCAGTTATTCAGCAGGAAAACTACCACTGTCAAAGTGCTGATTAGGGTAGCTCTCAAAGCTTGCGCGAGCGCGCTTGTGGGTAATCTGGGGGGGCCACGACCTGTCTGCTTTAGCCGACATGGTGTTAACGCTTTGGTTCTTCACGACTAAATGTAACTCGTTGCACACACACGTGAGCAACCACTCACAGCTACACTACAAACGGACCCGACAATTGGTCTTCTGATTCGTGTATCCTATGAATGACGGTTGAGTCTTCCTTAACTTAACCCCTTTCTTTCCCTCCCGTAAACTCCCTGCGCAAATGCCTACTGTATTAGTGAAAGGACAAACATGATTCGCCAAGGCCCGGAAACCCGCATTCGTCACATCTTAAACAAACTAGCATCCGCAACACTTCGAATGTGGCGAGAGATCCTCAATGGAGCCACGAATTCACGGAGGTGAACGTTATATGTAATCAAAGGGTAGCTGTGTCCAGCCCCGGTAGTTGTCATGAGCCTAAATCGTAAATTCAGTTTGCCCAGCGTCATACGTCGGATTAGAGAATCCTATCCGTTGG')        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transcribing DNA into RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CAGCUUGUCACAGCGACAAUUCUUUUAAGUGAAACAGGGUUUCGGACAUCAAUUCUACGGAUCAAAUUCCUAACGCGCGCCGGCUCGCUUUUUGACUGCAUGGCAUCUGCAUCUGAAGUCUUGCCGUGGGGUGCUUGUCAAUAGAAUUUGAAUGAAUGGUGUGAGCUGCCUUGGUGGCUGGUUCACUGCUACCGCCUCCUAUAUCGCUGAAAUUGAUCGCCCUUUUUAUCGGAACUGCUACGCUCUCGAUUAACGUGGUUACAACACGAGUAGUGAUCUAAAAGCCGCACGCGGGAAGGACCGGUCAAGGGUAAGUUAGUAACGGUCCAAACAGUGGUUUCUCGCACUAGGAUUAGUGGCAACUAAGCAGUGUGUACAACGCCACAGCCAUUGUGUGGACUAAGUUAUGCAGCAGUCGGGUAAUUGGUCCCAAAGGGUAACCAUUUCACUACGACGUUGUUGGAGAAGGUACUAGUCGAGUAAUAUUGUGGCAUUAGCUUUGGAUAAGGACAAACUUUUUGAUAUGCCGUCAAGGUAUCUUAGCCGCCGUAGGAACGCAAUACCCUUCCAGAUCUAUCCUUCCUGCUUGAUGCGGUUAAUUCACGCUUCAUAGUACGACAUACAUCAACUCAGUCAUAUCCUGAGUUACAUGGGGGACUACUCGCGGCUGUGGGCAACGAAUGGUGUGCAGCUAAAGUCACGCGACAUAUAAGAAGGUGCCAGGAUAAAUGUCGGGUCAACUCCGUCAGGACGUACUAUUCACCGGGACGAGCAGGGAUACGUGAUUACUACCCAGUCCAAGUCCGUUAGAACUGCUAGUGAGGGCCGAGUCAACCCCUGAACAGGCCGACGCUGUUCAUACCUUGGUCUAAUAAGUACUAUACACGCGAAUAUGUGCUUUCAGCACUUAGAUCGUCAUCACCCGGCGACUUUAACGAUAUUA'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def transcribing_dna_rna(dna_string):\n",
    "    list_dna = list(dna_string)\n",
    "    for n, x in enumerate(list_dna):\n",
    "        if x == 'T':\n",
    "            list_dna[n] = 'U'\n",
    "    return ''.join(list_dna)\n",
    "\n",
    "dna_string = 'CAGCTTGTCACAGCGACAATTCTTTTAAGTGAAACAGGGTTTCGGACATCAATTCTACGGATCAAATTCCTAACGCGCGCCGGCTCGCTTTTTGACTGCATGGCATCTGCATCTGAAGTCTTGCCGTGGGGTGCTTGTCAATAGAATTTGAATGAATGGTGTGAGCTGCCTTGGTGGCTGGTTCACTGCTACCGCCTCCTATATCGCTGAAATTGATCGCCCTTTTTATCGGAACTGCTACGCTCTCGATTAACGTGGTTACAACACGAGTAGTGATCTAAAAGCCGCACGCGGGAAGGACCGGTCAAGGGTAAGTTAGTAACGGTCCAAACAGTGGTTTCTCGCACTAGGATTAGTGGCAACTAAGCAGTGTGTACAACGCCACAGCCATTGTGTGGACTAAGTTATGCAGCAGTCGGGTAATTGGTCCCAAAGGGTAACCATTTCACTACGACGTTGTTGGAGAAGGTACTAGTCGAGTAATATTGTGGCATTAGCTTTGGATAAGGACAAACTTTTTGATATGCCGTCAAGGTATCTTAGCCGCCGTAGGAACGCAATACCCTTCCAGATCTATCCTTCCTGCTTGATGCGGTTAATTCACGCTTCATAGTACGACATACATCAACTCAGTCATATCCTGAGTTACATGGGGGACTACTCGCGGCTGTGGGCAACGAATGGTGTGCAGCTAAAGTCACGCGACATATAAGAAGGTGCCAGGATAAATGTCGGGTCAACTCCGTCAGGACGTACTATTCACCGGGACGAGCAGGGATACGTGATTACTACCCAGTCCAAGTCCGTTAGAACTGCTAGTGAGGGCCGAGTCAACCCCTGAACAGGCCGACGCTGTTCATACCTTGGTCTAATAAGTACTATACACGCGAATATGTGCTTTCAGCACTTAGATCGTCATCACCCGGCGACTTTAACGATATTA'\n",
    "transcribing_dna_rna(dna_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Complementing a Strand of DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CAAGCGTGATGTCGACAGCAAGTATAAGTTGATTGAGTAGATTTTAGACATAAAATCCGGTTGCAGAAGCTGCGTTGGCATAGTTGAACACCATATAGTTGGAAACTCGTAATAGACGTGAGCCTCGTGCCCCGGCTAATGTCAAGAACTGTAACTAACGTTGGCCAAGGTATCTCATATCACGCAGTCGAGAAATGACCTACTAGTAGTCTCCGAACCCAAAAAAGCGGATAGGTCACGCCTAGCTATGAGACGGCGTGTGAGAAGAGGCGCTTTAAAAGCTCTTACGGCACCGTGGTTGAAAACCGTGGCCGAGTCCTCGAAACAGACCGACCGCGATGGCATCGTATTAACTGAAATCATCTTTAGATCGCAAAGCCCATATTGACTTGGATTGCTTATAATTGTTTGATGCTCGACTACGGCTAACCTGAAACGCATGTCCTCCAGACTCAGGGCCATAGGAGAATAGACTGTGCATGTCTCACGGAAGTCCTCACCTAATAATTCTCTGCCCTGACATAAACACACTGGACCTAGTATGTTCACCAGGTTATTTACCAAGAACCTAACGAAACGAGAGACCGAGGATCCTTTCCATTCTCTGTGTAAGGCCGTGAATAAGCGGAATAAATCAGACTTCTAGTATGGGGAATCGACACGTCCCCCCATTGGGTAACTGCAATTTCCAAGTGGTGGAAGGCCATTATAATGCCCAGGCCGCCCAGGATAGTTATCTATAATTCCCGCTTAAGACACGGTTCTTCGAGGGAGGGAAGAGGCGCTATTAGTTTTGTCCGGCATGATTCGGTGGGAATTAGCTTCGGACATAACAAACCGTAACTTACGTTCGGTCGGAGTCTAGCAGAGATGCGTTGAGAAAAAGTACTGCGGGTAGGTCAGGAACCAGTACTGTGCTAAATTAGATCCTGGACCCATAACCTAAAC'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def complementing_strand(dna_string):\n",
    "    list_dna = list(dna_string)\n",
    "    for n, x in enumerate(list_dna):\n",
    "        if x == 'C':\n",
    "            list_dna[n] = 'G'\n",
    "        elif x == 'G':\n",
    "            list_dna[n] = 'C'\n",
    "        elif x == 'T':\n",
    "            list_dna[n] = 'A'\n",
    "        elif x == 'A':\n",
    "            list_dna[n] = 'T'\n",
    "    return ''.join(reversed(list_dna))\n",
    "\n",
    "dna_string = 'GTTTAGGTTATGGGTCCAGGATCTAATTTAGCACAGTACTGGTTCCTGACCTACCCGCAGTACTTTTTCTCAACGCATCTCTGCTAGACTCCGACCGAACGTAAGTTACGGTTTGTTATGTCCGAAGCTAATTCCCACCGAATCATGCCGGACAAAACTAATAGCGCCTCTTCCCTCCCTCGAAGAACCGTGTCTTAAGCGGGAATTATAGATAACTATCCTGGGCGGCCTGGGCATTATAATGGCCTTCCACCACTTGGAAATTGCAGTTACCCAATGGGGGGACGTGTCGATTCCCCATACTAGAAGTCTGATTTATTCCGCTTATTCACGGCCTTACACAGAGAATGGAAAGGATCCTCGGTCTCTCGTTTCGTTAGGTTCTTGGTAAATAACCTGGTGAACATACTAGGTCCAGTGTGTTTATGTCAGGGCAGAGAATTATTAGGTGAGGACTTCCGTGAGACATGCACAGTCTATTCTCCTATGGCCCTGAGTCTGGAGGACATGCGTTTCAGGTTAGCCGTAGTCGAGCATCAAACAATTATAAGCAATCCAAGTCAATATGGGCTTTGCGATCTAAAGATGATTTCAGTTAATACGATGCCATCGCGGTCGGTCTGTTTCGAGGACTCGGCCACGGTTTTCAACCACGGTGCCGTAAGAGCTTTTAAAGCGCCTCTTCTCACACGCCGTCTCATAGCTAGGCGTGACCTATCCGCTTTTTTGGGTTCGGAGACTACTAGTAGGTCATTTCTCGACTGCGTGATATGAGATACCTTGGCCAACGTTAGTTACAGTTCTTGACATTAGCCGGGGCACGAGGCTCACGTCTATTACGAGTTTCCAACTATATGGTGTTCAACTATGCCAACGCAGCTTCTGCAACCGGATTTTATGTCTAAAATCTACTCAATCAACTTATACTTGCTGTCGACATCACGCTTG'\n",
    "complementing_strand(dna_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Translating RNA into Protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "condon_table= {\n",
    "    'UUU': 'F',\n",
    "    'CUU': 'L',\n",
    "    'AUU': 'I',\n",
    "    'GUU': 'V',\n",
    "    'UUC': 'F',\n",
    "    'CUC': 'L',\n",
    "    'AUC': 'I',\n",
    "    'GUC': 'V',\n",
    "    'UUA': 'L',\n",
    "    'CUA': 'L',\n",
    "    'AUA': 'I',\n",
    "    'GUA': 'V',\n",
    "    'UUG': 'L',\n",
    "    'CUG': 'L',\n",
    "    'AUG': 'M',\n",
    "    'GUG': 'V',\n",
    "    'UCU': 'S',\n",
    "    'CCU': 'P',\n",
    "    'ACU': 'T',\n",
    "    'GCU': 'A',\n",
    "    'UCC': 'S',\n",
    "    'CCC': 'P',\n",
    "    'ACC': 'T',\n",
    "    'GCC': 'A',\n",
    "    'UCA': 'S',\n",
    "    'CCA': 'P',\n",
    "    'ACA': 'T',\n",
    "    'GCA': 'A',\n",
    "    'UCG': 'S',\n",
    "    'CCG': 'P',\n",
    "    'ACG': 'T',\n",
    "    'GCG': 'A',\n",
    "    'UAU': 'Y',\n",
    "    'CAU': 'H',\n",
    "    'AAU': 'N',\n",
    "    'GAU': 'D',\n",
    "    'UAC': 'Y',\n",
    "    'CAC': 'H',\n",
    "    'AAC': 'N',\n",
    "    'GAC': 'D',\n",
    "    'UAA': 'Stop',\n",
    "    'CAA': 'Q',\n",
    "    'AAA': 'K',\n",
    "    'GAA': 'E',\n",
    "    'UAG': 'Stop',\n",
    "    'CAG': 'Q',\n",
    "    'AAG': 'K',\n",
    "    'GAG': 'E',\n",
    "    'UGU': 'C',\n",
    "    'CGU': 'R',\n",
    "    'AGU': 'S',\n",
    "    'GGU': 'G',\n",
    "    'UGC': 'C',\n",
    "    'CGC': 'R',\n",
    "    'AGC': 'S',\n",
    "    'GGC': 'G',\n",
    "    'UGA': 'Stop',\n",
    "    'CGA': 'R',\n",
    "    'AGA': 'R',\n",
    "    'GGA': 'G',\n",
    "    'UGG': 'W',\n",
    "    'CGG': 'R',\n",
    "    'AGG': 'R',\n",
    "    'GGG': 'G'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MCANYGSLGSINRARSASYILFETALKNFRYVTASSGPFGWAKDPSCMSHAHPPFKPLSFSDFLRLPSRCLICTLRLGSRAFQKGATPKTPVWLSPNRFVDRLPNTGNQPHTLVAHLKQQIKRLNQVILAAIALFYDKHILETLPDYPLTLLNERQTGKSPPGMPRSILRRCPSSGCVYSRPGHVKNTTRMYHLRFRLTRTIWTRTTNPLACLRIECYSVDVRQATFACSRVHLCKTSTRERVFPSACGQRLPATIDCNGLNTRLGIPKLPPLCHQLPYSKSLRPYFNAWYSPESGTRVGRIHFVYFCFPAFEHLLYRQILLRLFMTHVKSIIMCLYTGRCELTPTVEQFHATPRLTATPIDFISPPLWGFFDRAGDPTACGHCSISTKAYHGVGARHWPGTTRSFETKSTYKTPLALCVVCPGLGIAAGMSGGGTCLRAAPMFPRRTRASLADCTSCPQRISIKRITTDSRKMDVAQGTYTQDGFLFSCAGSRAVSPESLRSSEKCATFFKLTNEFIQEKPRKGIKPVDLALRPDVFSCSWSGLHVSAPYGYLGLMKPRTLGAGRRDICVYAWSCLHASHRPRLPIKMTTLSEVLGMRPLTSVRTAGQVVESGTGRRCSHTSSPRYFVLVLRVDNVHNRTHLAQVRSRNSPKLSWRYSLRTVTVQGRCAQSIESMQTKRRSWTFTNAIYELVIRVSAGLASLLFSRNCGRMLHQSVIETYLLCCVSTGGRMIYLRINSYRLCGRDYWHLRSSWRTAINSCLTKHGRLSCNNDTVAGTPWGTFQSRGPMEHHLRGAALQLMQLNSVWAGTNDRAVMKRTLLTRTPLRDESASERFTYFGSLCNSMVPLRAHRRGEEISTLQGSRLTNPDGTQSLLSETASGAVGRNRSDGLYGDKEPTVLNYGRHRGEIIFVRTSLRNQGALTDSVTPIIVQPSSAYLAVFSALHVKIWRGPTDFIQPQCLTLEAGNLNGLAFLNHWVAKQLPVAQSQSSLFQKNMRAFLGKDSCVELARWEWRLIRSHTFKLILSSYRFYQQLPTPAVFRKELGMRSNAPHLLVIQIARRCRPLNRKPLWDSGSYAENGEQRNEEKKLFRRTINLSLLQLCHRPMRISSPVAVSPFLLYCVRHVWPELNKITHFGSIHSRTCRILLHLMTRPTPTGTKSDRSSSRRAQVTCRGISISCGGTNQLTEQLAECVERRSLLRSNDFRFIAMPESNNLWVLYMPATRVLRNRLLYAVLERLRGGRILDAIQGRVVGQGLRSRTRWGRLGAPCMTLWLNRTRCTTRLGCPRVSSEVRSRTCPLLIPLWGRDEALRLPHHPTHFDHAMYTKSAYSYSLAVLRAFRPNLAYCRSELKLINALGTTGCEGSDKIWFLRFCSNRFKCRGRTQPITTVVSSGEVPQSAFQGKGDFVLSWGQTSSTWRTRLQLSLQVTRSYVQSPLETFLMGSGSEMEADFNESDLEILAMRNRNLQADSGATWAPGVFYLLDGPMPTNTLRPNVDLDHMERVRYRTVSFLCRTSRQRHFARSLRSCTTPGRPSSTLSLARNMPTPVQTAKWAYVGFILLPSSGREQCALLAEPERAMYTYTRHQKWTRRQRFVPILATCRGWPLGQRRANIIAFVASRREWHSFRGHENVSTRHKGRPTRLKADCCGPRGRGIKPHQWFQLFLAHIVSARLRDVLRFQLWLYWVCSETALSILFIIKVRGGRPPLCPQCMSFQLELLRGYKTLAAAHAAMYMVSTQRRGSHGQPYFVKYPCLRDRNCSICLGWVDYMSYQTKPQICLRTSLPTTLTWSAGGLVPVGQRIRLLIDVSSLSSKCSSVESDPTCVSRWGNVPAEAYDRRRAGTGTTRFLTVSRCSTEWRCWYDGSDLTVVIIYKQGLIGPRVGTHPKYYVPCPFIEPNHGVLTTCTRLDQSDRKVAESETTAFFTPVDAASIQKEDHRETSGGVECAPSGMTRQHASNMMLYRLMVYFSSRITWKKSNRDVNRPTEKRQAGSIQNMGAIVTLARDRLWTSRVAHKSTSLPTKMHLYVSLDVSPTQTQQPRCLTGCGWGRNSLGGFSKCSTFQSNDVKGQGTTKIFRGLHTLFRCPYRPVNLRPDHKSPRCRSPVSMRDHRMLPVECKTSLNVIYLVSKPSGDLTVTVHSSKRVWRVDRTSILNRCRAYTGMVLFDVASQSRFERSIRGTVNPWLNSRQCMYCVPKTVKRSDMIVRTFWMASTHAYLCALVQPLIVKHQNHAGLNPQRRYVRLANYASLRSRIDDHRGSLRVRATCVREIDGSSNRTTDLLELRATAHGHQTGNCAPRWHLPLVQLIARHKGESTAELYICKQGLPIATIALRTVVRCILAAVRRNRELLLIMNWLVKGNAHREIHQITYGSLRIPNFIPECRVIESCTHMQTSLHVTVASTKLERYNSCRSVVWTSTSLEKLGSMMRTALPCAKLQVQVEPLRILAKIAYLILRVNTPALPPTIRTFDLRTFVIFDVSELESPKKLSFMHYPEADFKAVRWLADFSPVEKSSATKMVSPGAAAGASRSEIPPVFCKLRVIPVLYQTSQESSESSYLHIPMSAGLRCGAELWVYVSRVGHDDCNVKTASRYSLRSSSFRPTQSRGNPEEIYSILPALVVDYVQGAILLAVAAVLRYYQAVSNERLSVVNIWARPSRSGRPAAVYEALCVVMHSCRIAPCFGTERRQHLIQHSTATAKMAKVLTQYILNFESGFALRFTKPSRPEELILQQPNALTNGITFNQYDTPAVLLFSTRCVLVELAASQMDLRALEQYVMDSSHWPPLLAHYLDPVGISAAASRLEVSIQSLSRVVHGGNPLGGTALAFASIPTPVVNYLSDEAGHCALRQAPNPYCGVRVGWNNSLLLSPKQILLYIRH'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def chunks(l, n):\n",
    "    for i in range(0, len(l), n):\n",
    "        yield l[i:i + n]\n",
    "\n",
    "def translating_rna_protein(condon_table, rna_string):\n",
    "    \n",
    "    chunkers = list(chunks(list(rna_string), 3))\n",
    "    chunkers = [''.join(chunk) for chunk in chunkers]\n",
    "#     proteins = [ for chunk in chunkers in chunk == ]\n",
    "    output = []\n",
    "    for chunk in chunkers:\n",
    "        for protein in condon_table.keys():\n",
    "            if chunk == protein and condon_table[protein] != 'Stop':\n",
    "                output.append(condon_table[protein])\n",
    "\n",
    "    return ''.join(output)\n",
    "\n",
    "rna_string = 'AUGUGUGCAAACUAUGGCAGCCUAGGUUCUAUCAACCGUGCCCGGUCGGCUUCCUAUAUCCUCUUUGAAACAGCGUUAAAGAAUUUCCGUUACGUAACCGCCUCAUCUGGGCCUUUCGGGUGGGCUAAAGAUCCCUCAUGUAUGAGCCACGCACAUCCACCCUUUAAACCACUAAGCUUCAGUGACUUUUUGAGAUUGCCCUCACGUUGCCUGAUCUGUACUCUAAGGUUGGGUUCACGGGCCUUUCAAAAGGGCGCUACACCCAAAACGCCAGUGUGGCUCUCGCCUAAUAGAUUCGUGGACCGGUUGCCAAACACGGGGAACCAGCCUCACACACUAGUAGCCCACCUCAAACAACAAAUAAAACGGCUGAAUCAGGUUAUACUCGCGGCGAUUGCCCUGUUUUACGAUAAGCACAUAUUGGAAACGCUCCCAGACUACCCUUUAACACUGCUUAAUGAGCGACAGACUGGAAAAUCGCCACCCGGUAUGCCUAGAUCCAUUCUGAGGAGGUGCCCAUCUUCGGGGUGCGUCUACAGUAGGCCCGGGCACGUCAAAAACACUACCCGUAUGUAUCACCUCCGAUUCCGGUUAACGAGGACCAUAUGGACUAGGACCACAAAUCCUUUAGCCUGUCUGCGCAUAGAAUGUUAUUCUGUUGAUGUCAGACAAGCAACGUUCGCGUGCAGCCGCGUUCAUCUUUGUAAGACGUCAACAAGAGAGCGCGUGUUUCCAUCAGCAUGCGGGCAGCGCCUCCCAGCAACGAUCGACUGUAAUGGCCUGAACACACGGCUCGGUAUCCCCAAACUCCCUCCUCUAUGUCACCAGCUCCCAUACAGUAAAUCACUUCGUCCCUACUUUAAUGCCUGGUACAGUCCCGAAUCGGGCACAAGAGUGGGACGUAUACAUUUUGUCUAUUUCUGCUUUCCCGCCUUCGAGCACCUCCUAUACAGGCAAAUUCUACUUCGCCUUUUUAUGACGCAUGUAAAAUCAAUCAUUAUGUGCCUGUAUACAGGUCGAUGUGAAUUGACUCCCACCGUGGAGCAAUUUCACGCAACUCCGCGUCUUACCGCUACGCCUAUCGAUUUCAUCAGUCCGCCCCUUUGGGGUUUCUUCGAUCGGGCAGGAGAUCCCACUGCUUGUGGCCAUUGCAGCAUCUCUACCAAGGCUUAUCAUGGGGUAGGGGCCCGGCAUUGGCCAGGCACCACGCGAAGUUUUGAAACGAAGUCAACCUAUAAGACACCGCUGGCGUUAUGCGUCGUGUGUCCGGGUUUAGGAAUCGCGGCCGGUAUGUCGGGGGGGGGGACGUGUCUACGCGCCGCCCCCAUGUUUCCAAGGCGGACCCGAGCAUCUCUGGCAGACUGCACAAGCUGCCCGCAACGUAUCAGUAUAAAACGUAUUACAACGGAUAGUCGAAAGAUGGAUGUGGCACAAGGCACAUAUACGCAGGACGGGUUCCUGUUCUCCUGUGCAGGUUCUAGAGCGGUGAGCCCAGAGUCCCUAAGGUCUAGCGAGAAAUGCGCAACCUUCUUCAAGUUGACAAAUGAAUUCAUCCAGGAGAAACCCCGGAAGGGCAUUAAGCCGGUUGACUUGGCUUUGAGACCGGAUGUCUUCUCGUGUUCCUGGAGCGGUUUGCACGUUUCCGCGCCCUAUGGUUACCUAGGUCUAAUGAAACCGCGGACCCUGGGGGCUGGUCGUCGCGACAUUUGUGUUUAUGCGUGGUCUUGUCUCCAUGCAAGUCACCGGCCCCGUUUGCCAAUAAAGAUGACAACUCUAAGUGAAGUACUAGGAAUGCGCCCGUUGACUUCGGUACGCACCGCCGGGCAAGUUGUGGAAUCUGGCACAGGUCGACGAUGUAGCCACACCUCGAGUCCAAGGUACUUUGUCCUUGUGUUGAGAGUAGACAACGUGCAUAACAGGACCCAUUUGGCACAGGUGCGCUCGAGGAAUAGCCCAAAACUGAGUUGGAGGUAUAGCCUACGAACUGUCACGGUCCAGGGUAGAUGCGCCCAAAGCAUAGAAAGUAUGCAGACCAAACGAAGGUCCUGGACUUUCACGAACGCUAUCUACGAGCUCGUGAUACGUGUGUCUGCCGGCUUGGCUUCUUUAUUAUUCUCAAGAAAUUGUGGCCGUAUGUUACAUCAGAGUGUUAUCGAAACCUAUCUACUAUGUUGUGUAUCAACUGGAGGUCGCAUGAUAUACCUUCGGAUAAACUCCUACAGGCUCUGUGGGCGAGAUUAUUGGCACCUCCGUUCCUCAUGGCGAACGGCUAUAAACUCUUGCCUAACAAAACAUGGGCGCUUGUCGUGCAACAAUGAUACCGUCGCGGGUACGCCGUGGGGCACCUUUCAAUCCCGAGGUCCUAUGGAGCAUCAUCUACGUGGGGCUGCGCUGCAGCUAAUGCAACUAAACUCUGUGUGGGCCGGUACUAAUGACCGCGCAGUGAUGAAGCGUACGCUGCUAACUCGCACCCCCCUAAGAGAUGAAUCAGCGUCGGAAAGAUUCACCUAUUUUGGUUCGUUGUGCAACUCCAUGGUACCUCUCCGGGCGCAUCGCCGUGGAGAAGAAAUUUCUACCUUACAGGGCUCUAGAUUAACGAACCCUGAUGGUACACAAUCCUUGCUCAGCGAAACGGCCUCAGGUGCGGUAGGUAGGAAUCGGUCUGACGGGCUAUACGGAGAUAAAGAACCAACGGUUCUGAAUUACGGCCGACAUCGAGGUGAGAUCAUAUUUGUGAGGACGUCGCUACGGAAUCAAGGCGCUCUAACCGACUCUGUAACACCCAUCAUCGUCCAACCGAGUAGCGCUUAUUUAGCGGUCUUCUCUGCACUUCACGUAAAAAUUUGGCGUGGUCCAACCGAUUUCAUUCAACCCCAGUGUCUCACGUUGGAGGCAGGCAAUUUAAAUGGCCUUGCAUUCCUAAAUCAUUGGGUUGCCAAACAGUUGCCUGUUGCCCAGUCCCAGUCAUCGCUCUUUCAGAAGAAUAUGCGCGCCUUUCUUGGCAAGGAUUCGUGUGUGGAACUGGCCCGGUGGGAGUGGCGGCUCAUAAGGUCACACACCUUUAAGUUAAUUCUUUCCUCAUACAGAUUCUACCAACAACUUCCCACACCCGCGGUCUUCAGGAAAGAGUUAGGAAUGCGAAGUAACGCGCCGCAUCUCCUCGUCAUUCAGAUAGCCCGUCGUUGUAGACCUCUAAACCGCAAGCCGUUAUGGGACUCAGGAAGCUAUGCUGAGAACGGUGAACAGAGAAACGAAGAGAAAAAACUAUUUCGAAGAACGAUCAAUUUGUCUCUGCUGCAACUUUGUCACAGACCCAUGAGGAUCUCAAGUCCAGUAGCUGUCUCACCUUUCUUGUUGUAUUGUGUACGACACGUUUGGCCCGAGCUGAACAAAAUUACUCAUUUCGGUUCCAUUCAUUCCCGGACCUGCCGCAUUUUAUUGCACCUGAUGACCAGACCCACACCUACAGGUACCAAAUCCGACAGAUCGAGUAGCAGACGCGCUCAAGUAACCUGUAGAGGUAUAAGUAUUUCGUGUGGCGGCACCAACCAGCUGACGGAGCAGUUGGCGGAGUGUGUUGAACGGCGGAGUCUUCUGCGUAGCAACGAUUUCAGGUUCAUAGCUAUGCCAGAAAGCAACAACCUCUGGGUUUUAUAUAUGCCCGCGACACGUGUCCUUAGAAACCGACUCCUUUACGCGGUCUUAGAAAGGCUUAGGGGAGGGAGAAUCUUGGAUGCUAUCCAAGGACGCGUAGUCGGACAAGGGUUGCGAUCGCGCACCCGCUGGGGGCGACUCGGCGCCCCCUGUAUGACUCUAUGGUUAAAUCGAACCCGUUGUACGACGCGUCUUGGCUGUCCCCGCGUAUCCUCUGAAGUUCGAAGUCGCACUUGCCCGUUGCUCAUACCAUUGUGGGGCCGUGAUGAGGCACUACGCCUCCCACACCACCCCACACAUUUUGACCAUGCUAUGUACACGAAGUCUGCCUAUUCAUAUUCCCUUGCCGUUCUCAGGGCUUUCCGGCCGAACUUGGCAUAUUGCCGUAGCGAGUUAAAACUUAUUAACGCCCUUGGUACAACCGGAUGCGAAGGGUCGGACAAGAUUUGGUUCCUCAGGUUCUGUUCGAAUCGUUUCAAAUGUAGGGGCCGGACACAACCUAUCACUACCGUAGUGUCUAGUGGAGAAGUACCGCAGUCCGCAUUUCAGGGAAAAGGCGACUUUGUAUUGUCAUGGGGACAGACAAGCUCGACUUGGCGCACGAGGCUACAAUUGAGCUUACAAGUCACGAGGAGUUAUGUACAAUCACCUCUCGAAACGUUCCUGAUGGGAAGUGGGUCGGAAAUGGAAGCUGAUUUCAAUGAGUCGGAUUUAGAGAUCUUAGCCAUGCGUAACAGAAACCUUCAGGCUGACUCAGGCGCUACAUGGGCCCCUGGCGUCUUUUACCUAUUGGACGGGCCGAUGCCGACAAAUACGCUACGACCUAACGUGGACUUGGAUCAUAUGGAACGAGUGAGAUAUCGAACAGUGUCCUUCCUAUGCCGUACGUCACGCCAACGUCACUUUGCUAGGUCAUUGCGCAGUUGUACAACGCCGGGAAGGCCUUCAAGCACCUUAUCCCUAGCCAGAAAUAUGCCUACACCAGUGCAAACAGCUAAGUGGGCGUAUGUCGGUUUCAUAUUACUCCCCAGUUCAGGGCGUGAGCAAUGCGCUUUGUUAGCUGAACCUGAGCGCGCUAUGUACACAUACACUCGCCAUCAAAAAUGGACUAGACGACAACGCUUCGUUCCUAUCUUGGCAACCUGUCGAGGUUGGCCCCUUGGCCAGAGACGAGCGAACAUCAUUGCGUUCGUCGCUUCUCGACGCGAAUGGCAUAGCUUUAGGGGUCAUGAGAAUGUUUCUACGAGACACAAAGGGCGACCUACACGUCUAAAAGCAGAUUGUUGUGGGCCCCGUGGCCGAGGCAUAAAACCACAUCAAUGGUUUCAACUGUUCCUUGCUCAUAUAGUCAGCGCACGUCUGCGUGAUGUGUUGCGCUUCCAACUGUGGCUCUACUGGGUCUGCAGUGAAACGGCGCUGUCAAUCCUAUUCAUUAUCAAAGUUCGGGGCGGUAGGCCUCCAUUGUGCCCUCAAUGUAUGAGUUUCCAGCUGGAACUUUUACGAGGUUAUAAGACAUUAGCAGCUGCACACGCAGCAAUGUACAUGGUAAGUACACAGCGCCGGGGCAGUCACGGUCAACCGUACUUUGUCAAGUAUCCGUGCUUAAGGGACCGUAACUGCUCUAUCUGCUUGGGAUGGGUGGAUUACAUGUCCUACCAGACCAAACCGCAGAUUUGCCUUCGAACCUCUCUGCCGACUACGUUGACUUGGUCCGCGGGUGGACUCGUCCCCGUCGGCCAACGAAUACGUUUACUGAUUGAUGUGUCCUCCCUGUCAAGCAAGUGCAGCUCCGUUGAGUCAGAUCCCACAUGCGUGUCACGGUGGGGCAACGUUCCAGCAGAGGCGUAUGACCGUCGACGGGCUGGAACGGGAACUACCCGCUUUCUUACUGUAUCACGAUGCAGUACUGAAUGGCGAUGUUGGUAUGAUGGUUCCGACCUUACAGUUGUGAUCAUUUACAAGCAAGGCUUAAUUGGUCCUCGAGUCGGCACUCACCCGAAGUAUUAUGUGCCCUGUCCCUUUAUCGAGCCCAAUCAUGGGGUGUUGACGACUUGCACACGCCUGGACCAGAGCGACCGCAAAGUUGCAGAGUCGGAAACAACAGCGUUUUUUACACCAGUUGAUGCAGCAUCCAUACAAAAAGAAGAUCAUCGAGAGACCUCCGGGGGCGUAGAAUGUGCUCCGAGUGGGAUGACGCGUCAGCACGCAUCAAACAUGAUGUUGUAUCGACUGAUGGUAUACUUUUCAUCGAGAAUAACAUGGAAGAAGUCAAACCGCGACGUGAACCGCCCAACCGAAAAGAGACAGGCCGGAAGCAUCCAGAACAUGGGGGCUAUAGUCACCUUGGCAAGAGAUCGUCUUUGGACAUCGAGAGUCGCCCACAAGUCAACUAGCCUGCCCACGAAAAUGCACCUAUAUGUCAGCUUAGACGUUUCGCCAACGCAGACCCAGCAACCGAGGUGCCUGACAGGGUGCGGAUGGGGGAGGAAUUCACUCGGGGGGUUCAGUAAGUGUUCAACCUUUCAGUCCAACGAUGUGAAGGGGCAAGGUACGACUAAGAUCUUCCGGGGUCUUCAUACGCUUUUUAGGUGUCCCUACCGUCCAGUAAAUCUACGGCCCGACCAUAAAAGUCCGCGGUGUCGAUCCCCGGUCAGCAUGAGAGAUCACCGCAUGCUUCCCGUGGAAUGUAAGACUUCGCUAAACGUCAUCUAUCUGGUAUCAAAACCCUCCGGAGAUUUGACAGUCACGGUACAUAGUUCCAAAAGGGUAUGGCGGGUAGACCGCACCAGCAUUUUGAACAGAUGCCGAGCGUAUACCGGUAUGGUGCUUUUCGACGUCGCCUCGCAGAGCCGGUUCGAGCGCAGCAUUAGGGGGACAGUUAACCCUUGGCUCAAUAGUCGACAAUGUAUGUACUGCGUCCCAAAGACAGUUAAGCGCUCGGACAUGAUUGUGCGGACUUUUUGGAUGGCCUCGACUCACGCAUACUUGUGUGCUUUAGUGCAACCAUUGAUUGUAAAGCACCAAAACCAUGCCGGUCUAAACCCUCAGCGGCGUUAUGUGAGGUUAGCUAAUUACGCUUCCCUCCGUAGCCGAAUCGAUGACCACAGAGGUUCUCUGAGAGUCAGGGCCACCUGUGUACGAGAAAUUGAUGGUAGUUCCAACCGGACGACCGAUCUUCUUGAGCUCCGCGCGACAGCGCAUGGCCAUCAGACCGGAAAUUGUGCGCCGAGAUGGCAUCUACCACUAGUACAACUUAUAGCUAGACAUAAGGGCGAAUCGACGGCUGAACUAUAUAUUUGUAAACAGGGACUUCCCAUAGCCACCAUUGCCUUGCGCACGGUCGUUCGUUGCAUAUUAGCUGCCGUUCGGCGUAAUCGAGAACUGCUGCUAAUUAUGAACUGGCUGGUUAAAGGCAAUGCACAUCGUGAGAUUCAUCAGAUUACUUAUGGUUCUCUGAGGAUCCCAAACUUUAUUCCUGAGUGCAGAGUGAUUGAGUCUUGCACCCACAUGCAGACAUCUCUCCACGUGACGGUGGCGUCGACGAAGUUAGAGCGUUAUAACAGCUGUCGGAGCGUUGUCUGGACCUCUACAAGUUUGGAGAAGUUAGGCAGUAUGAUGCGAACUGCCCUUCCUUGUGCUAAGUUACAGGUGCAGGUUGAACCCUUGCGCAUACUGGCGAAAAUUGCCUACCUAAUCCUCCGGGUGAAUACACCCGCCCUACCACCAACCAUACGUACUUUCGACCUGCGAACCUUUGUAAUCUUUGAUGUCUCUGAAUUAGAGUCGCCCAAGAAGUUGAGUUUCAUGCACUAUCCCGAGGCCGACUUCAAAGCGGUAAGGUGGUUAGCCGACUUCUCACCAGUAGAAAAGUCAUCAGCGACGAAGAUGGUCUCGCCUGGUGCGGCUGCCGGCGCAAGCCGAUCAGAAAUACCACCGGUGUUUUGCAAGCUUAGGGUGAUACCCGUGCUAUACCAAACAAGUCAAGAAUCGAGUGAAUCUUCAUACUUGCAUAUCCCUAUGUCUGCAGGCUUGCGGUGCGGAGCAGAGCUGUGGGUAUACGUAUCACGAGUGGGACACGAUGACUGUAAUGUCAAAACGGCAUCAAGGUACUCUUUACGAUCUUCGUCUUUCCGUCCAACCCAAUCCCGUGGAAAUCCCGAAGAAAUAUAUAGCAUUUUACCCGCACUAGUCGUGGACUACGUCCAAGGAGCGAUACUCUUGGCCGUCGCCGCGGUGCUUCGAUAUUAUCAGGCAGUGUCCAAUGAACGUUUGAGCGUCGUGAAUAUUUGGGCCCGACCCAGCCGCUCUGGCAGGCCGGCGGCGGUCUAUGAAGCUUUAUGCGUCGUCAUGCACUCCUGCCGUAUUGCUCCUUGUUUUGGGACUGAACGUCGCCAGCAUCUUAUUCAGCAUAGUACAGCCACCGCGAAGAUGGCAAAGGUCCUGACCCAGUACAUCUUAAACUUCGAAUCUGGGUUUGCAUUAAGAUUCACCAAACCAAGUCGGCCGGAGGAAUUAAUACUCCAACAGCCUAAUGCCCUGACCAAUGGGAUUACAUUCAACCAAUACGACACGCCAGCGGUGCUGUUGUUCAGUACGCGAUGCGUGCUUGUUGAAUUGGCAGCCUCCCAAAUGGAUCUCCGAGCGUUAGAACAAUAUGUGAUGGAUUCUAGUCAUUGGCCUCCGCUGUUGGCACACUAUUUGGAUCCUGUAGGCAUUUCAGCUGCGGCGUCGAGGCUGGAGGUGAGCAUACAAUCAUUAUCGCGCGUCGUACAUGGGGGUAAUCCGCUUGGCGGCACAGCGUUAGCCUUCGCCUCUAUACCAACACCAGUAGUCAAUUACCUAAGUGAUGAGGCGGGGCACUGUGCUCUCCGUCAAGCACCUAACCCUUAUUGUGGUGUGCGAGUAGGGUGGAACAAUUCUCUACUGCUCUCUCCUAAGCAGAUCCUAUUGUAUAUACGACACUGA'\n",
    "translating_rna_protein(condon_table, rna_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculationg protein mass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "monoisotopic_mass_table = {\n",
    "'A': 71.03711,\n",
    "'C': 103.00919,\n",
    "'D': 115.02694,\n",
    "'E': 129.04259,\n",
    "'F': 147.06841,\n",
    "'G': 57.02146,\n",
    "'H': 137.05891,\n",
    "'I': 113.08406,\n",
    "'K': 128.09496,\n",
    "'L': 113.08406,\n",
    "'M': 131.04049,\n",
    "'N': 114.04293,\n",
    "'P': 97.05276,\n",
    "'Q': 128.05858,\n",
    "'R': 156.10111,\n",
    "'S': 87.03203,\n",
    "'T': 101.04768,\n",
    "'V': 99.06841,\n",
    "'W': 186.07931,\n",
    "'Y': 163.06333 \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "96830.31"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def calculating_mass(protein_string):\n",
    "    list_protein = list(protein_string)\n",
    "    mass = 0\n",
    "    for x in list_protein:\n",
    "        mass += monoisotopic_mass_table[x]\n",
    "    return round(mass, 3)\n",
    "\n",
    "protein_string = 'IAEACHPAKNTVTLVAYSQYFAKWKNRISWWSMTVVSVIYVDDSESSLQWHTHCRLIAEAWYSMRIMWCVRTIGVDCCEPHVTIDVRCKTMAESATKWRYKPWAGGKWDDMMKFEMSGMGPRMWIPVRDWGGFNDWQHKYKDKIRWCCWCWKWILPDCHYLMQHTDTQCPYVQVHQMFQNTPFPCMPTGFPFLFSKRFKVNLRLYGYVPRTHMLISCKGRSFMMFMLPGQVATTPIGGPVTEREYSHTECWDKAMSATFDMMERVTLLIAHDCMPAYLYGLKMIHLPVRRSKDGRWGCPWAHVVHSINPCMTPPWLITHTVPKIWRKVSHPVPMRQHQTVHVALGTCQAKSYEVWFVEHCPHLIQSDIILRYTNIVVFLTLTACFPSPPVTGPTMSNPMFADSQQKTIFCSAAHYWNCIPGHQYKAVFDRQGPDWSLTIYCVSTCICFRFRDPYNYEYKACSAGNKSMAHDGHKDHNKDLYGNVINVHIEWIAFDCYIEQTIHVGLCIPNVDQNCEIVCDDIHWWGSCSLQCEPDERMIHWWVVTIKMNDGKTLCEGHIDCKDNIMMEAEYMQSIYLWGKIRQTNSYIRWQWKWGIIGIAVKNIPPEKFISAVVPRSLYCGGCNVYSAFRLDCWEELVEQAPHTPFLPHEIASGMPTPILLCLQVQVGGKLRGPLWLYKQTCWMDDFSHADMGMREHVDRFDKMIFSVCDKTNCGGKGNQSGKCCNTDGSMYSYKCFKNNIQLTWAVGIMIRVLTLMHHRVPDLRQMVFKFYECRGHWRRHNWEHVNEQEPGWFHDIRESSNGHIFQPCHDKWRVFRSFEENDYNR'\n",
    "calculating_mass(protein_string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing GC Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_3778\n",
      "53.66705471478463\n"
     ]
    }
   ],
   "source": [
    "from numpy import argmax\n",
    "file_dir = 'rosalind_gc.txt'\n",
    "\n",
    "ids = []\n",
    "dna_strings = []\n",
    "\n",
    "with open(file_dir, 'r') as file:\n",
    "    dna_strings = []\n",
    "    chunk = []\n",
    "    for line in file.readlines():\n",
    "        if line.startswith('>'):\n",
    "            ids.append(line[1:-1])\n",
    "            dna_strings.append(chunk)\n",
    "            chunk = []\n",
    "        elif line == '\\n':\n",
    "            dna_strings.append(chunk)\n",
    "        else:\n",
    "            chunk.append(line)\n",
    "\n",
    "    clean_dnas = []\n",
    "    for dna in dna_strings[1:]:\n",
    "        sub_clean_dnas = []\n",
    "        for x in dna:\n",
    "            if str(x).endswith('\\n'):\n",
    "                sub_clean_dnas.append(x[:-1])\n",
    "        clean_dnas.append(sub_clean_dnas)\n",
    "        sub_clean_dnas = []\n",
    "        \n",
    "    cleaner_dna = []\n",
    "    for clean_dna in clean_dnas:\n",
    "        cleaner_dna.append(''.join(clean_dna))\n",
    "        \n",
    "    percentage_list = []    \n",
    "    for _list in cleaner_dna:\n",
    "        percentage = (_list.count('G') + _list.count('C')) / len(_list)\n",
    "        percentage_list.append(percentage)\n",
    "            \n",
    "print(ids[argmax(percentage_list)]) \n",
    "print(max(percentage_list)*100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### using biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TACTACCTGTGAATGGACGACCAAGGGCTATGGGGAAACTATAAGTCCCTGTAAACGCACACTTAGCCCTAAGTAAGCACAGGTTTCTCCTAGATCGGGCAAATATTCGCGCTAGAAAAACCCTGAGCTCATGTTACTCTATCATAAGTCACCTGCACCGGATCCCGGACTTTTTTCCAATCTCTATGGATAGCTGCGATTTGTCGCAGCTGGAGTCGGCCCATACCGTCACGAGTGGGTATCCACATGCGCCTTGGTTTCATGCGTAATTCGACGGCCGAGCTATGACGTCCGCGGGGCCAGGCCGTGGCGATAAGGTATCGGTAGTATCTGGGTGACCAATCTAGCACGACACACAACTATTGACATTGTCACCCGAGCTGTAGCATGGGCTTAAGAGCCTAATGGTGCGCGTACCCTTTATAGCCTTACCTGTCGTAGAGCCCGCAATAGTCTGATCACCCGCTTATTCTGCGCTAGAGGAAGAATGCCGGTATCCGAGATATCATTATGCTAGATAGCCGAAGGCATAGAACACGGGTCAACCAGCAATCATGTAGCCATCAGAGGATTGAGTTATCGTGTTTCCAGCCCGGCTTGACAACATGAGAGTAAATCGAGAACTGCAGTTTAAACCGACCGGAGTGCATTAAACCGCAACGGTAATCGTTCTTCTGGCGAAAAACTGGTCCGCGTCCTACACTGTCGCGCGGTGGCCTAGGGGCAAACCAAGAGGTCGTCCCAGAGATAGCCGGTCAGTTATGTAGAAGCTTGTCACGATAAGGTCGGCGGGCGTCCAATGG'"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "with open('rosalind_gc.txt') as fasta_file:  # Will close handle cleanly\n",
    "    identifiers = []\n",
    "    sequences = []\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        identifiers.append(seq_record.id)\n",
    "        sequences.append(seq_record.seq)\n",
    "sq = str(sequences[0])\n",
    "sq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mendel's First Law"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8377024722932651"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.special import comb\n",
    "\n",
    "def calculate_probability(k, m, n):\n",
    "    total_pop = k + m + n \n",
    "    total_combos = comb(total_pop, 2)\n",
    "    valid_combos = comb(k, 2) + k*m + k*n + .5*m*n + .75*comb(m, 2)\n",
    "    probability = valid_combos / total_combos\n",
    "    return probability\n",
    "\n",
    "calculate_probability(30, 22, 17)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating expected offspring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "148987.5\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "def calc_expect_offspring(population):\n",
    "    nums = [int(x) for x in population.split(' ')]\n",
    "    es = [0.75 * nums[3], 0.5 * nums[4]]\n",
    "    for i in range(3):\n",
    "        es.append(nums[i])\n",
    "    print(sum(es)*2)\n",
    "    \n",
    "population = '16681 19731 17083 16277 17582 16306'\n",
    "calc_expect_offspring(population)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding a Motif in DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41 57 91 107 132 149 177 184 191 207 214 252 330 399 435 487 504 511 599 656 696 724 751 793 800 807 823 846 861 939 947 957\n"
     ]
    }
   ],
   "source": [
    "def finding_motif(sstring, tstring):\n",
    "    result = []\n",
    "    tlen = len(tstring)\n",
    "    slen = len(sstring)\n",
    "    for i in range(slen - tlen + 1):\n",
    "        if sstring[i:i+tlen] == tstring:\n",
    "            result.append(i+1)        \n",
    "    print(' '.join(map(lambda x: str(x), result)))\n",
    "    \n",
    "sstring = 'ACTACCTATCCGCCTATCCTCGCAAAGAACAGCCTATCCTCCTATCCCCTATCCCTCCTATCCCCCTATCCGCCTATCCGTCCTATCCCACCTATCCCCTATCCTTCCTATCCCCTATCCACCTATCCAAACCTATCCCCCCTATCCACCTATCCCCCCTATCCCAACCCTATCCACCTATCCCCTATCCCCTATCCCCTATCCTACCTATCCCCTATCCCCGCCTATCCGCGACCTATCCGCCTTATCACCCTATCCCCTATCCCACGGAACCTATCCAGGACCTATCCACCTATCCACCTATCCTGCAACCCTATCCATCCTATCCGCCTATCCCCTATCCACCTCCCTATCCAGCCTATCCGGATTCCTATCCGAAACCTATCCTTCCCTATCCGCCTATCCCCTTAATCCCTATCCCGTATCCTATCCGACCTATCCCCTATCCGCTTGTCCCCCTATCCGTCCTATCCGACCTATCCAAACCCTATCCCCTCTTAAAGCCTATCCCCTATCCCCCTATCCACCTATCCTCCTATCCATGCACGTGACCTATCCAAGAAACCTATCCAGCCTATCCACGTGGGGATCGAAGTGTCCTATCCCCTATCCTGGTCAACCTATCCAAACCTATCCTCCTATCCCTCCTATCCCGCCTATCCCCCTATCCCGTTGTCCTATCCTCAACCTATCCGCCTATCCCCTATCCTACCTATCCACCCTCCTATCCCCTATCCAGATCCTATCCGCCCTATCCCCATCCTATCCAAACCTATCCTTCCGTAATTATACCCTATCCCCTATCCCCTATCCCCTATCCGTCCTATCCCCTATCCAGTCTGCGCCCTATCCCCTATCCACCTATCCCCTATCCACCCTATCCAACCGGCCTATCCTCGCCCCCTATCCACCCTATCCGACACCGTGAAGTTCATAGACCTATCCCCCTATCCCCACCTATCCCCTATCCGGCCTATCC'\n",
    "tstring = 'CCTATCCCC'\n",
    "finding_motif(sstring, tstring)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RNA Splicing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MWASGRGRIRTVNTPYAGADRKSMHNRVLKTFSPAIDTTKHPPGSPWRTAFIPALGISLNLPGTFHVVVRGLASVCCLIAWHPIHITRYIWGASRRNNNVEELTLSYPKFPAGLRLAHLVVGPFSVLPNTNSTPPQPSAHHSTATDFRTTLRRENICPRDGCSRSTLCRFPSTCMYARPLFGVW*'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import re\n",
    "\n",
    "with open('rosalind_splc.txt') as fasta_file:\n",
    "    identifiers = []\n",
    "    sequences = []\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "        identifiers.append(seq_record.id)\n",
    "        sequences.append(seq_record.seq)\n",
    "for intron in sequences[1:]:\n",
    "    sequences[0] = Seq(re.sub(str(intron), '', str(sequences[0])))\n",
    "\n",
    "proteins = sequences[0].transcribe().translate()        \n",
    "str(proteins)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counting Point Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "433\n"
     ]
    }
   ],
   "source": [
    "def counting_p_mut(dna1, dna2):\n",
    "    hamming_distance = 0\n",
    "    for x, y in zip(dna1, dna2):\n",
    "        if x != y:\n",
    "            hamming_distance += 1\n",
    "    print(hamming_distance)\n",
    "    \n",
    "dna1 = 'CACCGTCCCTCGCATCAACCAAGAGTGAGAACTTGGGATACTCAGAGTGGATGTCAGCAGGTTATGTCGTTTCCTGCTTCCGCATTGTTAAGCGCGCTCTCGGCCGGATGAGCAATCCGGCGAAGCGCGATGAATGTCGTTTTAGGCCTATAGGTGGACGTAGCCCGACAGCGATCACAGAACCCCTGTGACTCGAAGAAAGAGCATGTCATGGGGGTCCGAGACCTGGATTTGTGGCGCGAGAATACCTTACTAGGCCCCTAATGCCAATGTGATGCCAGCACTCTTATGTTTTATCTTAGGTCGCGCACCAATTTATTACAGCTTGGTCGAATACTCATTTCCCGCCTTTTTACCAAATTGACAGGAATCTCGGGTTACAACCCCCAGGCGCACATAATATTACGGTGTTCACCATAGGTCGGCATCCTTTCCACCGTCGTTGGAGACTGGAAGGAGCTACTTGGTTATAGACGCCAACAAATTTGTATATGTCTGGATTGCATTCAGGTGAAGGAAGGGGTATAGGATACCGAGCCATTCCGCCTGGAGGTACTTAGCGTCGGAGGCCAAAATACGTCTAGGAATAGTGCTTGACTGCAGTTGAACGTCTACAACGTGTGCCGAGTGCCGTAAGTCCTACTCCTTCGGGGTAGAGACCAGTCTATGAGCGTCCTGCACACTGCAAGCAACGCTGAGGAAAGTGTGCTCAGGTCTAGATTCTGTTGCTTAAGCAATCCGATAATTATGTGGCTTAGGACATGTGTTCCTCTGGCAAATGAACCATAATAAGTTTGGGCAGCCCACAGGGCGCTCCCATCTGCCGTGTCGTACTGCGATGAGTGCATGATCGACATTACCTCAAGGAGTCCCCAGTGCCTAGCCTCAGAGGTTTGCTCTCGCGT'\n",
    "dna2 = 'CGGACTACCGCCGAGAATTGGACACTGAGAGAGTGGGCGGGTACGAGGATTTTGATCAACTTTATGTCGAGCCCTGTATCTAAATCCTTTGAGCCGCGCTGTGCCGTAACGGCTATCGTGCTAATCGTGCTGAAAGGGGCTTTATACAAAAAGGTTGGACAAGCCCGTAAGCGTGTGAAGAAACGTCCGGTGAACAAGATTATCTAGTAGGTATCAAACCGAGACCTGGTCTCGTTTCAAGAGATGTAAAAGCTCGGCGACCTATGACGGTTGGAATCGAACCGGCCTGTGGTGAGGGGGAGTTCGCTCGGAAGCTTCCTCGAACCGAGTAGAATATCGATATCAAGCACGCAGGCCAAACTGACTGGCCTCGCGAGATACTACCTCCAGGCGGGCATTACATTAAGGGTTTCAACATTGTTCGTTCGCAGGGCCGATTCCCTTCGACACTGGAAAGAGGTATTCCTTTATACAGGCGTCCATATCTATATATATCTGGGCTTGATACAGAAGAACGATCTGTCCTAGGCTTCGAAATGTCTGCCTATCGTGGTAAAGAGAGCTGCAGTTCAAGATGCCTCCCCGATAGTTGATTGCTAGTAATGCGAAGGATACCACTTAGTATGAAGGCCGTATGGCACTTTAGTCCCGCGTATAGTTCAGTCCTTTGGCGGCCTGGATCATAAGCTATGTGATGGAGAATGTGGGTGCACTTTTAGCCACTAGTCTTCTAGCATTCAGCAAAGGATGAGGCCAATGGGAGGTGGACCTCTTTTCGTTGCATGCTATTTAGAATCGGCTCTCGACAACAAACTGCCCTCTGAGGCGTAGGAATTCGCTTGCAGCACCACCCGAATTAATTAGGGGAGTCCCCAACGCCTATTCTGAGACCTTTAACATCGAGC'\n",
    "counting_p_mut(dna1, dna2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transitions and transversions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.11340206185567\n"
     ]
    }
   ],
   "source": [
    "with open('rosalind_tran.txt', 'r') as f:\n",
    "    raw_data = list(SeqIO.parse(f, 'fasta'))\n",
    "\n",
    "seq1 = raw_data[0].seq\n",
    "seq2 = raw_data[1].seq\n",
    "\n",
    "transitions = 0\n",
    "transversions = float(0)\n",
    "for nt in range(len(seq1)):\n",
    "    if seq1[nt] == seq2[nt]:\n",
    "        continue\n",
    "    elif seq1[nt] == \"A\" and (seq2[nt] == \"C\" or seq2[nt] == \"T\"):\n",
    "        transversions += 1\n",
    "    elif seq1[nt] == \"G\" and (seq2[nt] == \"C\" or seq2[nt] == \"T\"):\n",
    "        transversions += 1\n",
    "    elif seq1[nt] == \"C\" and (seq2[nt] == \"A\" or seq2[nt] == \"G\"):\n",
    "        transversions += 1\n",
    "    elif seq1[nt] == \"T\" and (seq2[nt] == \"A\" or seq2[nt] == \"G\"):\n",
    "        transversions += 1\n",
    "    else:\n",
    "        transitions += 1\n",
    "\n",
    "print(transitions/transversions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find shared motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'5 6 13 17 19 21 23 30 36 38 40 46 49 50 57 58 66 67 78 82 83 88 92 95 97 104 112 114 116 123 124 127 128 135 145 150 153 155 156 161 162 163 164 167 169 170 182 188 193 195 198 202 203 204 206 209 210 211 212 213 218 226 229 233 234 238 239 242 259 260 266 268 270 278 279 282 300 306 308 312 316 318 329 332 334 339 343 352 356 360 369'"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "with open('rosalind_sseq.txt') as fasta_file:\n",
    "    sequences = list(SeqIO.parse(fasta_file, 'fasta'))\n",
    "\n",
    "seq_list = []\n",
    "\n",
    "for sequence in sequences:\n",
    "    seq_list.append(sequence.seq)\n",
    "                     \n",
    "motif = []\n",
    "prev_num = 0\n",
    "for sequence in seq_list[1:]:\n",
    "    for char in sequence:\n",
    "#         print(seq_list[0])\n",
    "        num = seq_list[0].find(char)\n",
    "        motif.append(num + prev_num + 1)\n",
    "        seq_list[0] = seq_list[0][num + 1:]\n",
    "        prev_num = num  + prev_num + 1\n",
    "        \n",
    "motif_str = ''\n",
    "for num in motif:\n",
    "    motif_str = motif_str + ' ' + str(num)\n",
    "motif_str[1:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Locating restriction sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 4\n",
      "39 4\n",
      "54 4\n",
      "55 4\n",
      "56 4\n",
      "81 4\n",
      "117 4\n",
      "131 4\n",
      "141 4\n",
      "157 4\n",
      "191 4\n",
      "196 4\n",
      "202 4\n",
      "210 4\n",
      "238 4\n",
      "246 4\n",
      "303 4\n",
      "305 4\n",
      "307 4\n",
      "321 4\n",
      "335 4\n",
      "338 4\n",
      "346 4\n",
      "356 4\n",
      "365 4\n",
      "381 4\n",
      "403 4\n",
      "406 4\n",
      "409 4\n",
      "415 4\n",
      "419 4\n",
      "467 4\n",
      "473 4\n",
      "508 4\n",
      "510 4\n",
      "526 4\n",
      "534 4\n",
      "593 4\n",
      "614 4\n",
      "622 4\n",
      "638 4\n",
      "641 4\n",
      "656 4\n",
      "671 4\n",
      "682 4\n",
      "684 4\n",
      "687 4\n",
      "702 4\n",
      "704 4\n",
      "792 4\n",
      "796 4\n",
      "810 4\n",
      "811 4\n",
      "54 6\n",
      "209 6\n",
      "245 6\n",
      "304 6\n",
      "334 6\n",
      "405 6\n",
      "414 6\n",
      "533 6\n",
      "592 6\n",
      "621 6\n",
      "683 6\n",
      "53 8\n",
      "303 8\n",
      "404 8\n",
      "620 8\n",
      "52 10\n",
      "403 10\n",
      "619 10\n",
      "402 12\n"
     ]
    }
   ],
   "source": [
    "def rev_complement(text):\n",
    "    rev = ''\n",
    "    for c in text:\n",
    "        if c == 'A':\n",
    "            rev = 'T' + rev\n",
    "        elif c == 'C':\n",
    "            rev = 'G' + rev\n",
    "        elif c == 'G':\n",
    "            rev = 'C' + rev\n",
    "        elif c == 'T':\n",
    "            rev = 'A' + rev\n",
    "    return rev\n",
    "\n",
    "def is_rev_palindrome(chars):\n",
    "    return chars == rev_complement(chars)\n",
    "\n",
    "dna = 'ATATGGTATTGCCCGTCCAAGGCACGGATAACATAACTCTAGGGAGGCACCCTTATATAAGTGAAAGTCCTTCACCTATCTATACAAGAAGAGGTTTCAGAGAAGATTTCACAACCGATCTGCCATTCCCGATCTTACAGACGTGGCTGTGACACGATATGCTGTGAGCATAAAAGACCTGAATACTTCGTATAGAATTTCCTAGCAAACCGGTCAGCCGATAGTAAAGAAAAGCCACATGACACCATGGAAGTTGGTCTACAACTACTGACCTCAAAAGGCTCCTCTGTGGTGTCGCTCCAGTACGTACCTCGGGGGAAGATCACCGAGGGTTACGTATAACTAAGCTACCGCTCGCGGCTATTGCACCTCAGACTTATGGCCGCAAGCGTGTCGCCAACTTTAAGCTTAAAACTAGTGCATACGCCACCGAGGGAGGTCTAACTGGCGGGCTCTTCTTATTAGACCGGGTCCGGGCAATGTGGTTAGGAGGTGACCAATCATAGTTTAATTTGTTCAGGTATCGGCCACAGGATCCCCAGATAAATGCCGTGTGCCGCTGAGAGGTGTTGGCTTCCTCTGTAAACTGACCCGCGGCTACCCTAAACGCCGTCCGGTGAGCATGCTCCATTGAGGTTGCAGCTACATCCCCTGCGTACCAAAAAGTATTACGTTGAATCTGGCCGGCGCTGTCAGGGTGTTGCATGAGACACCGTTGGTGTCATTCTATTCTTCAGATAATCGTCACACGAAGCACCTTAGGATTCGTGAACGGATGTCTACAGAGACGGGATCACGTTCTCGTTGTCATATA'\n",
    "for l in range(4, 13):\n",
    "    for i in range(0, len(dna) - l + 1):\n",
    "        if is_rev_palindrome(dna[i:i+l]):\n",
    "            print(str(i+1) + ' ' + str(l))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
